feat: add health_generate route to openai serving

dsingal0 opened the pull request:
Adds /health_generate for kubernetes to detect when the runtime is hung so it can restart the container/pod even if the server /health returns a 200.
This is helpful in cases when a long context request from a user has caused the llm runtime to hang or die but the fastapi server is still running.

dsingal0 commented:
@kaiyux wdyt?

kaiyux commented:
> @kaiyux wdyt?

@dsingal0 Thanks for the suggestion. Some members in the team are on public holidays and will return next week, we will keep you posted.

At first glance I think we should be careful when introducing a new API that is not within OpenAI API scope, and I'll take a closer look on that.

LinPoly reviewed (commented):
For the added entrypoint function, it is acceptable to check if the runtime/executor/engine is alive. But similar logic can also be added to the client code instead of the server side, set a timeout for a simple request, then check if there is any response returned by the server and the response status. Do we have any reason to implement the logic in the server side? @dsingal0

LinPoly commented:
FYI: vLLM monitors executor health with daemon thread, it is more structured and reliable for me, but we need to implement similar checking from executor level. @kaiyux

LinPoly commented:
> @LinPoly
> 
> > re implementing the check server side vs client side
> > for deployment in kubernetes we need a health and liveness probe to detect when a pod is healthy and ready to be sent traffic during autoscaling and when its unhealthy and needs to be restarted. There is no client code to add this logic to in that case. Currently trtllm-serve exposes /health but that only checks if the server is up, not the health of the runtime. Would be great to have it at the executor/runtime level if possible

So you mean K8S is using server side health check for error detection and auto-scaling? If so I think it is reasonable to add such a WAR, we can work on a more structured solution afterwards. @kaiyux for opinions.

Superjomn commented on tensorrt_llm/serve/openai_server.py:
```diff
@@ -143,6 +144,41 @@ async def get_kv_cache_events(self) -> JSONResponse:
             pass
         return JSONResponse(content=events)
+    
+    async def health_generate(self) -> Response:
```
Currently, our `/health` endpoint is just a placeholder that always returns 200. If we implement it properly, returning an error code when there's an internal hang, would that address the motivation behind this WAR?

Alternatively, as a quick WAR, would it make sense to move `/health_generate` to `/health`, since they appear to serve the same purpose? Althrough we should have a more decent implementation in the future.

cc @kaiyux @penli9 for viz.

dsingal0 replied:
having /health return a non-200 response on internal error would be great.
Moving the /health_generate WAR to /health could work, but not everyone agrees with it because it introduces overhead. Eg if your health check is once every 10 seconds, you're generating 1 token and consuming one slot of the max_batch size every 10 seconds. For long running requests this can skew benchmarks numbers.

Superjomn replied:
Got it. We are going to introduce a heartbeat mechanism to detect and update the system's status, which should return by `/health`. `/health_generate` is fine as we have no such endpoint/functionality yet.

Superjomn commented:
@kaiyux @penli9 , as we have no such check yet, I think a `/health_generate` is reasonable; we can refine the implementation during further iterations.

kaiyux commented:
@dsingal0 Can you also help fix the DCO check? https://github.com/NVIDIA/TensorRT-LLM/pull/3856/checks?check_run_id=41920082141

> Summary
Commit sha: [2cd150e](https://github.com/NVIDIA/TensorRT-LLM/pull/3856/commits/2cd150ecb4163fef41f56c3864cc14db0c30decf), Author: Dhruv Singal, Committer: Dhruv Singal; The sign-off is missing.
Commit sha: [a12154a](https://github.com/NVIDIA/TensorRT-LLM/pull/3856/commits/a12154a74c3e0c6ac9e5f526be607b8d3f7bbad7), Author: Dhruv Singal, Committer: Dhruv Singal; The sign-off is missing.
Commit sha: [06f3d21](https://github.com/NVIDIA/TensorRT-LLM/pull/3856/commits/06f3d2106875133ce1264cf339482f95cd501c13), Author: Dhruv Singal, Committer: Dhruv Singal; The sign-off is missing.

You can follow the guidance here to sign off https://github.com/NVIDIA/TensorRT-LLM/blob/main/CONTRIBUTING.md#signing-your-work.

Please let us know for any questions, thanks!

dsingal0 commented:
@kaiyux done

kaiyux commented on tensorrt_llm/_torch/speculative/mtp.py:
```diff
@@ -689,7 +689,7 @@ def sample_and_accept_draft_tokens(
                                          dtype=torch.int,
                                          device=logits.device)
-
+        
```
Seems to be unintentional changes?

kaiyux commented on tensorrt_llm/serve/openai_server.py:
```diff
@@ -144,6 +157,40 @@ async def get_kv_cache_events(self) -> JSONResponse:
             pass
         return JSONResponse(content=events)
 
+    async def health_generate(self) -> Response:
```
Can you help add a test for this function? Please refer to https://github.com/NVIDIA/TensorRT-LLM/blob/main/tests/unittest/llmapi/apps/_test_llm_server.py#L28

dsingal0 commented:
@kaiyux I think rebasing to add signoff added those commits to the PR, can we squash on merge instead? I added the test and removed the mtp.py change.

dsingal0 commented:
alternatively, I could open a new PR

kaiyux commented:
> @kaiyux I think rebasing to add signoff added those commits to the PR, can we squash on merge instead? I added the test and removed the mtp.py change.

Thanks a lot! I think we should be fine merge this one since we will squash the commits when merge it, @chzblych do you think differently?

kaiyux commented:
/bot run

kaiyux commented:
/bot run

tensorrt-cicd commented:
[PR_Github #4705](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4705/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #4694](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4694/) [ run ] completed with state `ABORTED`

kaiyux commented:
@dsingal0 Can you help fix the failed test case?

```
[2025-05-09T14:06:26.867Z] ../../unittest/llmapi/apps/_test_llm_server.py .F....                    [100%]
[2025-05-09T14:06:26.867Z] 
[2025-05-09T14:06:26.867Z] =================================== FAILURES ===================================
[2025-05-09T14:06:26.867Z] _____________________________ test_health_generate _____________________________
[2025-05-09T14:06:26.867Z] 
[2025-05-09T14:06:26.867Z] client = <starlette.testclient.TestClient object at 0x7f632c50bf50>
[2025-05-09T14:06:26.867Z] 
[2025-05-09T14:06:26.867Z]     def test_health_generate(client):
[2025-05-09T14:06:26.867Z]         response = client.get("/health_generate")
[2025-05-09T14:06:26.867Z] >       assert response.status_code == 200
[2025-05-09T14:06:26.867Z] E       assert 404 == 200
[2025-05-09T14:06:26.867Z] E        +  where 404 = <Response [404 Not Found]>.status_code
[2025-05-09T14:06:26.867Z] 
[2025-05-09T14:06:26.867Z] ../../unittest/llmapi/apps/_test_llm_server.py:40: AssertionError
```

dsingal0 commented:
@kaiyux  I believe that fixed it, the test was failing since I had it in the llm api server, not the openai server.

kaiyux commented:
/bot run

tensorrt-cicd commented:
[PR_Github #4873](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4873/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #4873](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/4873/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #3531](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/3531/) completed with status: 'FAILURE'

kaiyux commented:
@dsingal0 Can you help fix the style check? : )  I would like to fix it but I noticed the PR is based on your main branch and I think I should not push something onto that.

kaiyux commented:
Hi @dsingal0 , I fixed the style check and cherry-picked the changes to https://github.com/NVIDIA/TensorRT-LLM/pull/4349, I marked you as co-author. Please let me know for any questions.

dsingal0 commented:
Amazing, thank you so much @kaiyux

kaiyux commented:
> Amazing, thank you so much @kaiyux

No problems, thanks a lot for the great support!
